### PR TITLE
raftstore: make sure PrepareFlashback will get the latest region meta

### LINKED ISSUE #13643: flashback ddl will hang when restart all pd nodes during flashback command
### What version of TiKV are you using?

Release Version:   6.4.0-alpha
Edition:           Community
Git Commit Hash:   ed64ed21cfa1c734191549a0db66986dab04f4bd
Git Commit Branch: master
UTC Build Time:    2022-10-18 09:38:46
### Steps to reproduce
1. execute flashback
2. during flashback, restart all pd nodes
### What did you expect?
after pd restart, flashback can execute successfully
### What did happened?
flashback ddl hang
```code 
[2022/10/20 21:08:51.094 +08:00] [INFO] [region_cache.go:1734] ["region epoch is ahead of tikv"] [error="region epoch is ahead of tikv. rpc ctx: region ID: 7714, meta: id:7714 start_key:\"t\\200\\000\\000\\000\\000\\000\\000h_r\\200\\000\\000\\000\\000\\0036\\332\" end_key:\"t\\200\\000\\000\\000\\000\\000\\000h_r\\200\\000\\000\\000\\000\\0037\\335\" region_epoch:<conf_ver:13 version:215 > peers:<id:7715 store_id:1 > peers:<id:7716 store_id:5 > peers:<id:7717 store_id:7 > , peer: id:7715 store_id:1 , addr: [172.16.6.47:2360](http://172.16.6.47:2360/), idx: 0, reqStoreType: TiKvOnly, runStoreType: tikv, currentRegions: [id:7714 start_key:\"t\\200\\000\\000\\000\\000\\000\\000\\377h_r\\200\\000\\000\\000\\000\\377\\0035!\\000\\000\\000\\000\\000\\372\" end_key:\"t\\200\\000\\000\\000\\000\\000\\000\\377h_r\\200\\000\\000\\000\\000\\377\\0037\\335\\000\\000\\000\\000\\000\\372\" region_epoch:<conf_ver:13 version:214 > peers:<id:7715 store_id:1 > peers:<id:7716 store_id:5 > peers:<id:7717 store_id:7 > ]"]
```

### PR DESCRIPTION
**Pull request by JmPotato**
Signed-off-by: JmPotato <ghzpotato@gmail.com>

### What is changed and how it works?

Issue Number: Close #13643.

What's Changed:

```commit-message
* Use `self.region` while executing `PrepareFlashback` command to ensure it gets the latest region meta.
* Check the epoch before executing the `PrepareFlashback`.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None.
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -2821,28 +2824,37 @@ where
...
             Ok(Some(s)) => s,
             _ => {
-                return Err(box_err!("failed to get region state of {}", region_id));
+                return Err(box_err!(
+                    "failed to get region local state of {}",
+                    region_id
+                ));
             }
         };
-        let is_in_flashback = req.get_cmd_type() == AdminCmdType::PrepareFlashback;
```
Comment (HuSharp) [inline components/raftstore/src/store/fsm/apply.rs:2833]: Why need to move this line?

File: components/raftstore/src/store/util.rs [hunk 2]
```diff
@@ -192,8 +192,12 @@ pub fn admin_cmd_epoch_lookup(admin_cmp_type: AdminCmdType) -> AdminCmdEpochStat
         AdminCmdType::RollbackMerge => AdminCmdEpochState::new(true, true, true, false),
         // Transfer leader
         AdminCmdType::TransferLeader => AdminCmdEpochState::new(true, true, false, false),
+        // PrepareFlashback could be committed successfully before a split being applied, so we need
+        // to check the epoch to make sure it's sent to a correct key range.
+        // NOTICE: FinishFlashback will never meet the epoch not match error since any scheduling
+        // before it's forbidden.
         AdminCmdType::PrepareFlashback | AdminCmdType::FinishFlashback => {
-            AdminCmdEpochState::new(false, false, false, false)
+            AdminCmdEpochState::new(true, true, false, false)
```
Comment (HuSharp) [inline components/raftstore/src/store/util.rs:200]: Do we need to add test in `test_check_region_epoch` to make sure `epoch.conf_version` was checked?

File: components/raftstore/src/store/fsm/apply.rs (hunk 1 above)
Comment (JmPotato) [reply components/raftstore/src/store/fsm/apply.rs:2833]: Just some tiny reorganization to make the code structure clear.

File: components/raftstore/src/store/util.rs (hunk 2 above)
Comment (JmPotato) [reply components/raftstore/src/store/util.rs:200]: It's covered by `test_prepare_flashback_after_split`.

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -2821,28 +2824,37 @@ where
     ) -> Result<(AdminResponse, ApplyResult<EK::Snapshot>)> {
         let region_id = self.region_id();
         let region_state_key = keys::region_state_key(region_id);
+        let is_in_flashback = req.get_cmd_type() == AdminCmdType::PrepareFlashback;
+        // We need to make sure the `RegionLocalState` read from the engine is latest,
+        // so the flush is needed before this function being called.
         let mut old_state = match ctx
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/apply.rs:2830]: no need to get it from disk, use `self.region`

File: components/raftstore/src/store/fsm/apply.rs [hunk 4]
```diff
@@ -758,7 +758,10 @@ fn should_write_to_engine(cmd: &RaftCmdRequest) -> bool {
             AdminCmdType::ComputeHash |
             // Merge needs to get the latest apply index.
             AdminCmdType::CommitMerge |
-            AdminCmdType::RollbackMerge => return true,
+            AdminCmdType::RollbackMerge |
+            // Flashback needs to get the latest region meta.
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/apply.rs:762]: no need now as we get the latest region meta from memory

Comment (JmPotato) [reply components/raftstore/src/store/fsm/apply.rs]: Fixed. PTAL.

**Comment by JmPotato** [comment]
/cc @Connor1996 @AndreMouche @HuSharp

PTAL again, thx!

**Review by Connor1996 — APPROVED** [review]
LGTM

**Comment by HuSharp** [comment]
LGTM

**Review by AndreMouche — APPROVED** [review]
LGTM

**Pull request merged**